lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -140,94 +145,229 @@
-        else:
-            installed_packages[package_details['name']].append(package_details)
-    return installed_packages
-
-
-# FIXME: add more listing methods
-def main():
-    global module
-    module = AnsibleModule(argument_spec=dict(manager=dict()), supports_check_mode=True)
-    manager = module.params['manager']
-    packages = {}
-    results = {}
+from ansible.module_utils.basic import AnsibleModule, get_all_subclasses
+from ansible.module_utils._text import to_native, to_text
+
+
+class PkgMgr:
+
+    def __init__(self, module):
+
+        self.m = module
+
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        raise NotImplementedError
+
+    def list_installed(self):
+        ''' This method should return a list of installed packages, each list item will be passed to get_package_details '''
+        raise NotImplementedError
+
+    def get_package_details(self, package):
+        ''' This takes a 'package' item and returns a dictionary with the package information, name and version are minimal requirements '''
+        raise NotImplementedError
+
+    def get_packages(self):
+        ''' Take all of the above and return a dictionary of lists of dictionaries (package = list of installed versions) '''
+
+        installed_packages = {}
+        for package in self.list_installed():
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.__class__.__name__.lower()
+            name = package_details['name']
+            if name not in installed_packages:
+                installed_packages[name] = [package_details]
+            else:
+                installed_packages[name].append(package_details)
+        return installed_packages
+
+
+class LibMgr(PkgMgr):
+
+    LIB = None
+
+    def __init__(self, module):
+
+        super(LibMgr, self).__init__(module)
+        self.lib = None
+
+    def test(self):
+        found = False
+        try:
+            self.lib = __import__(self.LIB)
+            found = True
+        except ImportError:
+            pass
+        return found
+
+
+class CLIMgr(PkgMgr):
+
+    CLI = None
+
+    def __init__(self, module):
+
+        super(CLIMgr, self).__init__(module)
+        self.cli = None
+
+    def test(self):
+        self.cli = self.m.get_bin_path(self.CLI, False)
+        return bool(self.cli)
+
+
+class APT(LibMgr):
+
+    LIB = 'apt'
+
+    def test(self):
+        works = super(APT, self).test()
+        if works:
+            self._cache = self.lib.Cache()
+        return works
+
+    def list_installed(self):
+        return [pk for pk in self._cache.keys() if self._cache[pk].is_installed]
+
+    def get_package_details(self, package):
+        ac_pkg = self._cache[package].installed
+        return dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, category=ac_pkg.section, origin=ac_pkg.origins[0].origin)
```

Comment:
MINOR: i think we decided to prefer literals over calling dict().

Reply:
mostly copy/paste from original, i'm fine either way

Reply:
let's stick to literals, please